### ISSUE TITLE
Package dependencies

### ISSUE THREAD

**Issue by sbcgua** [larshp/abapGit#2236]
Probably was discussed in some issues but I didn't find a dedicated one. Who is interested - join to discussion.

Goal: a package may have dependencies/prerequisites, let's mention them in `.abapgit`

Implementation ideas:
- dependencies must refer https git url + optional branch/tag. If we ever come to package management this will not conflict as packages will be probably centralized and have non-url-like name (like e.g. npm)
- mirrors are not considered or checked at this point, this is not the package management (yet)
- on install, AG should check if these urls already installed.
  - if yes, it should check same branch/tag and warn if not the same (not error, at the risk of user)
  - if no, user is notified and repos are installed separately
- after installation the repos live own lives, all the risk is on the user. Except:
  - deletion of a repo must check if it is a prerequisite for another one in the system
  - potentially on update AG can check if prerequisites were also updated since previoud update/install and inform user

- UI
  - prerequisite repos should have special flag which allows to hide/filter them from repo list. For convenience. It should be easy to remove the flag if the user want to start contributing/fixing. It is just for UX, no further restrictions.
  - editing dependencies allows to enter only those available on the system (arguable), branch/tag is freely choosable though (arguable)

**Comment by larshp** [comment]
sounds good, this is similar to my suggestion in https://github.com/larshp/abapGit/issues/217

suggest renaming the feature to "git dependencies" instead of "package dependencies", as package is a SAP object, which might have dependencies via package checks

UI: special flag, suggest leaving this out

Also some way for handling forks is needed, if someone clones a repo, it should recognize as the original already installed.

**Comment by sbcgua** [comment]
- renaming: agree, done
- UI flag: I'm sure it will be very useful. I have some ideas how to improve current repo listing, will PR with time. My current list of repos is longer than the screen - it is very inconvenient. But anyway the UI is at the end.
- Forks: hmmm, do you know how to find it out at github/gitlab and other supported providers ? That would be great but could be not easy to do (though maybe I'm just not in the subject ...)

**Comment by larshp** [comment]
another complication: if the repo is installed via ZIP, I would expect it to also recognize as installed

Been thinking of including eg. `zabapgit.pinf.xml`, but package interfaces does not work in ABAP cloud. Another approach is to add a key field in the .abapgit.xml file, but it would be difficult to maintain in a decentralized approach.

**Comment by larshp** [comment]
okay, so if we restrict the feature to online repos, and assume repositories are named uniquely, ie. abapGit is always abapGit, I guess it will work most of the time

**Comment by sbcgua** [comment]
yes, no way to auto install offline anyway. so dependencies should be just ignored for offline. And then reponame is relatively reliable factor.

**Comment by mbtools** [comment]
Lots of good ideas mentioned here already. Just offering some thoughts. I’m looking at challenging topics always from the end: what’s the ideal solution? For me it would be exactly what Node and npm do. So let’s assume we will have package.json in our repo to drive everything. No more abapgit.xml, no APACK. Whatever they contained or did needs to move into package.json. 

Only two mandatory fields: name and version. No other ids like urls or forks or telling where to find the repo. Dependencies are defined the same using name/version. 

Is the name unique? Not necessarily. But if you want to have something reusable you register centrally at npmjs. Dotabap.org should do the trick. 

I also like the optional fields like description, author, license, homepage, etc. abapgit is missing those. 

Abapgit.xml could easily move into a section of the json. Software component dependencies could actually be handled like other deps: “sap_basis/7.3.17” could do the trick (we might need to allow multiple entries for a “component”). 

It’s all declarative. Online or offline or urls or forks come in at a different time. And that leads me to the requirement that we need an installer (apm). I believe apm must be able to run without having an abapgit repo. We can develop it as part of abapgit but with the intention of being able to “abapmerge” a stand-alone version. Input is the zip format, not a repo clone. apm could get the zip online, but always installs like an offline scenario without (!) creating an abapgit repo. (One can easily create such repo if they link it to the sap package that was use by apm. )

I personally need apm to distribute apps I build (and will sell). It must be as easy to install itself as the standalone abapgit repo. And do just “one” thing: install or update abapgit zips including dependencies. The user shall provide all zips (or urls pointing to them if the system is connected to the net).
There are some other details to handle like target package, namespace, and transports but that’s already on my agenda.  

For this to become reality, I believe a refactor is required to separate repo and object/file layer ($abapgit_object). Apm would use apis of the object layer only: deserialize to install/update and delete to uninstall. Right now “repo” is input for the object layer but it needs to be the repo content (lt_tadir and lt_files/results). The split is not easy but doable. I will work on it.

To make it simpler for apm users, it would be nice to includea default target package and if necessary the namespace key in package.json  that could make  abapgit new repo easier too (create/pick package becomes optional). 

This leaves one big topic: how to handle multiple versions of the same repo. In git it’s a release tag/zip. And apm will take these zips as input based on which version the user wants to install. With things as they are only one version can exist in each system. 

But it’s possible to have many versions of the same code in a system... it just needs to be in different namespaces! So the killer feature will be to switch the namespace during the install. Compare it to putting a version of a node package into a specific folder. You can pick a different version for a different folder. The same will work with ddic and code in ABAP with namespaces. Going from y/z to namespace is a bit crazy but if we assume the source already has a namespace, then the switch becomes straight forward. 

Of course this limits reusable projects to objects that support namespaces but all the important part would work like Ddic and code. I call this an ABAP Library and it would make many people very happy if this became reality. 

Side note: abapgit has moved away significantly from “code” to include many non-core object types. Not sure this can or even should be stopped. However, I don’t see a problem limiting what can go into an ABAP library repo. It should be what’s most valuable for reuse: code and ddic. 

I have put together some slides with details which I will share in some other way since there are no attachments here. 
 
As you can tell, I see abapgit, repos, and git quite separate from apm. The former is essential for building and the later for the distribution. It’s really no different from the node/git world. The closer we mimick this, the more success we will have. 

I missed a ton for sure. Let’s keep the discussion going and build something incredible. Good fun ahead!

**Comment by flaiker** [comment]
> But it’s possible to have many versions of the same code in a system...  it just needs to be in different namespaces!

Maybe, I guess the namespace renaming tools help doing that (the open source ones / the obviously SAP internally existing one), though I doubt they are any near reliable enough to do it without user interaction. But the persistence layer seems like a nightmare. First of all DB table names already have super short names and if you require version 1.0.1, 1.0.2 and 1.0.3 of the same dependency you would have the DB tables 3 times? The 1.0.3 version does not have access to the other ones? Should it not? What happens if you want to upgrade a dependency? Does it migrate its 1.0.1 DB table into the existing 1.0.3 one (with conflicts???) or does it create 1.0.3b? What about the APIs? If the namespace is dynamic then all external usages need to be dynamic as well? Not all usages can be dynamic (views for example).

You can of course decide to ignore the persistence layer somewhat in the current open source ABAP world where most repos basically don't have it as they are BASIS tooling. But business applications in ABAP are to the core "just" CRUD applications with lots of customizing tables as well. I assume including the ones you want to package this way.

**Comment by mbtools** [comment]
The DB table exists only once per namespace. Same as with code. 

I put together an example of how the result should look:
1. An ABAP Library 
2. An app using the ABAP library 

See https://github.com/mbtools/abaplib-appdemo/blob/main/README.md

The library is at v3 and its table /libdemo/table has new country and city fields.
The app is still at v2 of the library and table /appdemo/table does not have country and city fields yet.

Typically, users only want the app. And developers want to test and distribute their app with a certain set of libraries (and update the libs only ondemand). That's why copying into the app namespace really makes sense. 

I built the example manually. But the idea is that apm takes a release ZIP of libdemo (https://github.com/mbtools/abaplib-libdemo/releases) and imports it into a package of the app namespace. There would be some additional steps needed like
renaming the files to the target namespace and replacing any /libdemo/ namespace found in the content of the files with /appdemo/. Both are pretty trivial to do. To update a library in an app, apm just repeats the process with a newer ZIP.

The trick is to have namespaces everywhere. Then switching them is easy.

**Comment by larshp** [comment]
heh, so I liked the title of this issue, because it was limited to just referring to git dependencies, but yea, it also needs to fit into a bigger picture with versioning, building, distribution etc.

renaming is trivial in some cases, when making assumptions regarding the code. 

If doing renaming, I think it should be done in the build step in the software lifecycle, not when installing the software, if using a git client it should take the remote and install, doing as few changes to it as possible.

versioning is not really a concept in ABAP, there is the latest version and that's it. If introducing versions, and two applications use the same library with database tables and transaction names, it will be very difficult to use IMHO. The author of the library would have to decide a new namespace for each version? 

Anyhow, the author of an application might choose to package the application along with a library, that the author has found to be "versionable", note that this does not impose responsibility to the library author or how/when/if to version recommended by the tools(eg abapGit).

**Comment by larshp** [comment]
Some more ideas:

The author of the library chooses to use the renaming approach. For each release create a `release/1` branch, this triggers a build on CI that will automatically rename all objects in that branch, creates tags and zip etc.

The author of the application can refer to the git dependency via tag/branch, change the git dependency to use a different tag/version. Rename everything again and package it into a zip if software distribution is via zip

**Comment by flaiker** [comment]
@mbtools I see, I didn't understand the part of copying the dependency into your own namespace at first. That is basically what abapGit did with ajson right? Comes with it's own caveats though. I guess it's the same discussion as package and dependency management on Linux with pros and cons on global "managed" dependencies by the OS in comparison to bundled dependencies by snap/flatpack/appimage/docker.

**Comment by mbtools** [comment]
Yes, it addresses exactly what we did getting ajson (and stringmap) added to abapGit. 

- imho, the library developer doesn't need versioning inside his system. it's provided by the remote git repo. all library development stays in one namespace (/libdemo/ in the example). 
- agree, abapGit is should not be involved in the renaming (I wasn't precise about that). the initial copy and renaming of a library for using it within an app is handled by apm. Updating all used libraries is again done by apm - you can call this a build process if you like.
- the library code becomes part of the app (in the app namespace but in a separate SAP package. of course, it is then included in the corresponding app abapGit repo.

Back to ajson. This is a perfect example. But it was a manual, one-time copy and renaming process to get it included in abapGit. How would we handle any updates? Diff doesn't even help. But with namespaces, you can automate it. Ajson would be developed in /ajson/ and abapgit in /abapgit/ namespace (just for the example!):
```
Library: /ajson/cl_ajson
App: /abapgit/cl_ajson
```
When a new version of ajson is released (update of package.json/release tag), you could run apm to copy it into /abapgit/cl_ajson. Run our tests. Done. 

apm should work independent from abapGit but the overlap is significant (mostly everything in object layer, HTTP, ZIP, and some other utility classes). So it makes sense to develop them together.

**Comment by larshp** [comment]
Status, since this issue got referred to,

[ajson](https://github.com/sbcgua/ajson) is mirrored into https://github.com/abapGit/ajson_mirror automatically including renaming, run `npm run mirror`, it is completely transparent to anyone if/when/how this happens, using existing tooling part of the git platform.

https://github.com/abapGit/abapGit/pull/4255 has been created manually, but when/if it gets working, this can be setup to run automatically using existing tooling as part of the git platform. 

All steps of this process are testable and transparent, all systems are the same, the developers will have identical code, no tricks. This also works in a decentral setup, most git platforms have automation(CI) that can perform Linux based steps.
